File: training/llmviz_train/mamba.py
"""Minimal Mamba reference implementation with activation recording.

This module is the golden source for the TypeScript port: every named
intermediate recorded here is compared against the TS re-implementation
at 1e-4 tolerance. Clarity therefore beats speed throughout — the
selective scan is a plain Python loop over time, one step per token.

Architecture (pre-norm residual stack):

    tokens -> embedding -> [x = x + block_i(rmsnorm_i(x))] * n_layer
           -> final_norm -> lm_head (untied) -> logits

MambaBlock (d_inner = expand * d_model):

    in_proj -> split (x_part, z)
    x_part -> depthwise causal conv1d -> SiLU
           -> x_proj -> split (dt, B, C)
           -> delta = softplus(dt_proj(dt))
           -> selective scan with A = -exp(A_log), D skip
    out = out_proj(y * silu(z))
"""

import math

import torch
import torch.nn.functional as F
from torch import nn


def _snapshot(tensor: torch.Tensor) -> torch.Tensor:
    """Detached float32 CPU copy of ``tensor`` for activation recording.

    The trailing clone matters: on a CPU float32 model, ``.to`` is a no-op
    that would alias live forward-pass storage, so later in-place ops could
    silently corrupt the snapshot.
    """
    return tensor.detach().to("cpu", torch.float32).clone()


class RMSNorm(nn.Module):
    """x * rsqrt(mean(x^2, dim=-1) + eps) * weight (no bias)."""

    def __init__(self, dim: int, eps: float = 1e-5):
        super().__init__()
        self.eps = eps
        self.weight = nn.Parameter(torch.ones(dim))

    def forward(self, x: torch.Tensor) -> torch.Tensor:
        return x * torch.rsqrt(x.pow(2).mean(dim=-1, keepdim=True) + self.eps) * self.weight


class MambaBlock(nn.Module):
    """One selective-state-space block (no residual; the Mamba module adds it)."""

    def __init__(self, d_model: int, d_state: int, d_conv: int, expand: int):
        super().__init__()
        self.d_model = d_model
        self.d_state = d_state
        self.d_conv = d_conv
        self.d_inner = expand * d_model
        self.dt_rank = math.ceil(d_model / 16)

        self.in_proj = nn.Linear(d_model, 2 * self.d_inner, bias=False)
        self.conv1d = nn.Conv1d(
            self.d_inner,
            self.d_inner,
            kernel_size=d_conv,
            groups=self.d_inner,
            padding=d_conv - 1,
            bias=True,
        )
        self.x_proj = nn.Linear(self.d_inner, self.dt_rank + 2 * d_state, bias=False)
        self.dt_proj = nn.Linear(self.dt_rank, self.d_inner, bias=True)

        # Standard Mamba dt bias init: dt ~ exp(U(log 1e-3, log 1e-1)),
        # bias = inverse-softplus(dt) = dt + log(-expm1(-dt)).
        dt = torch.exp(
            torch.rand(self.d_inner) * (math.log(1e-1) - math.log(1e-3)) + math.log(1e-3)
        ).clamp(min=1e-4)
        with torch.no_grad():
            self.dt_proj.bias.copy_(dt + torch.log(-torch.expm1(-dt)))

        # A_log init: each of the d_inner rows is log(1, 2, ..., d_state).
        a = torch.arange(1, d_state + 1, dtype=torch.float32).repeat(self.d_inner, 1)
        self.A_log = nn.Parameter(torch.log(a))
        self.D = nn.Parameter(torch.ones(self.d_inner))

        self.out_proj = nn.Linear(self.d_inner, d_model, bias=False)

    def forward(
        self, x: torch.Tensor, record: dict[str, torch.Tensor] | None = None, prefix: str = ""
    ) -> torch.Tensor:
        """x: (B, T, d_model) -> (B, T, d_model).

        When ``record`` is a dict, named intermediates for batch item 0 are
        stored into it (detached float32 CPU tensors) under ``prefix``.
        """
        B_, T, _ = x.shape

        def rec(name: str, tensor: torch.Tensor) -> None:
            if record is not None:
                record[prefix + name] = _snapshot(tensor)

        xz = self.in_proj(x)  # (B, T, 2*d_inner)
        rec("in_proj.out", xz[0])
        x_part, z = xz.chunk(2, dim=-1)  # each (B, T, d_inner)

        # Depthwise causal conv along time; padding=d_conv-1 makes the output
        # longer than T, so truncate back to the first T steps (causal).
        x_conv = self.conv1d(x_part.transpose(1, 2))[:, :, :T].transpose(1, 2)
        x_conv = F.silu(x_conv)  # (B, T, d_inner)
        rec("conv.out", x_conv[0])

        dbc = self.x_proj(x_conv)  # (B, T, dt_rank + 2*d_state)
        rec("x_proj.out", dbc[0])
        dt, B_mat, C_mat = torch.split(
            dbc, [self.dt_rank, self.d_state, self.d_state], dim=-1
        )
        delta = F.softplus(self.dt_proj(dt))  # (B, T, d_inner)
        rec("delta.out", delta[0])

        A = -torch.exp(self.A_log)  # (d_inner, d_state)

        # Selective scan, one token at a time.
        h = x.new_zeros(B_, self.d_inner, self.d_state)
        ys = []
        for t in range(T):
            delta_t = delta[:, t]  # (B, d_inner)
            B_t = B_mat[:, t]  # (B, d_state)
            C_t = C_mat[:, t]  # (B, d_state)
            x_t = x_conv[:, t]  # (B, d_inner)
            h = (
                torch.exp(delta_t[..., None] * A) * h
                + delta_t[..., None] * B_t[:, None, :] * x_t[..., None]
            )
            rec(f"ssm.h.t{t}", h[0])
            y_t = torch.einsum("bds,bs->bd", h, C_t) + self.D * x_t
            ys.append(y_t)
        y = torch.stack(ys, dim=1)  # (B, T, d_inner)
        rec("ssm.out", y[0])

        gated = y * F.silu(z)
        rec("gate.out", gated[0])

        out = self.out_proj(gated)
        rec("out_proj.out", out[0])
        return out


class Mamba(nn.Module):
    """Minimal Mamba language model over a small token vocabulary."""

    def __init__(self, cfg: dict, vocab_size: int):
        super().__init__()
        n_layer = cfg["n_layer"]
        d_model = cfg["d_model"]

        self.embedding = nn.Embedding(vocab_size, d_model)
        self.norms = nn.ModuleList(RMSNorm(d_model) for _ in range(n_layer))
        self.blocks = nn.ModuleList(
            MambaBlock(d_model, cfg["d_state"], cfg["d_conv"], cfg["expand"])
            for _ in range(n_layer)
        )
        self.final_norm = RMSNorm(d_model)
        self.lm_head = nn.Linear(d_model, vocab_size, bias=False)  # untied

    def forward(
        self, tokens: torch.Tensor, record: bool = False
    ) -> torch.Tensor | tuple[torch.Tensor, dict[str, torch.Tensor]]:
        """tokens: (B, T) long -> logits (B, T, vocab).

        With ``record=True``, also returns a dict of named activations for
        batch item 0 (detached float32 CPU tensors).
        """
        acts: dict[str, torch.Tensor] | None = {} if record else None

        def rec(name: str, tensor: torch.Tensor) -> None:
            if acts is not None:
                acts[name] = _snapshot(tensor)

        x = self.embedding(tokens)
        rec("embed.out", x[0])

        for i, (norm, block) in enumerate(zip(self.norms, self.blocks)):
            normed = norm(x)
            rec(f"layer{i}.norm.out", normed[0])
            x = x + block(normed, record=acts, prefix=f"layer{i}.")

        x = self.final_norm(x)
        rec("final_norm.out", x[0])

        logits = self.lm_head(x)
        rec("head.logits", logits[0])

        if acts is not None:
            return logits, acts
        return logits

File: src/compute/mamba.ts
/**
 * TS port of the Mamba forward pass in training/llmviz_train/mamba.py,
 * operation-for-operation, recording the same activation names. Gated by
 * mamba.golden.test.ts against PyTorch goldens at 1e-4.
 *
 * Architecture (pre-norm residual stack):
 *
 *     tokens -> embedding -> [x = x + block_i(rmsnorm_i(x))] * n_layer
 *            -> final_norm -> lm_head (untied) -> logits
 *
 * MambaBlock (d_inner = expand * d_model):
 *
 *     in_proj -> split (x_part, z)                       [x_part first]
 *     x_part -> depthwise causal conv1d -> SiLU
 *            -> x_proj -> split (dt, B, C)               [in that order]
 *            -> delta = softplus(dt_proj(dt))
 *            -> selective scan with A = -exp(A_log), D skip
 *     out = out_proj(y * silu(z))
 */

import { T } from "./tensor";
import {
  add,
  causalDepthwiseConv1d,
  embedding,
  exp,
  linear,
  mul,
  rmsnorm,
  scale,
  silu,
  softplus,
} from "./ops";
import { getTensor, type Manifest } from "./loader";
import type { Recorder } from "./recorder";
import { dimsFromManifest, sliceCols } from "./util";

/** Mamba dims as exported in manifest.json (dt_rank = ceil(d_model/16)). */
export interface MambaDims {
  n_layer: number;
  d_model: number;
  d_state: number;
  d_conv: number;
  expand: number;
  vocab_size: number;
  dt_rank: number;
}

const MAMBA_DIM_KEYS = [
  "n_layer",
  "d_model",
  "d_state",
  "d_conv",
  "expand",
  "vocab_size",
  "dt_rank",
] as const;

/** Validated MambaDims from a manifest (throws on wrong arch / missing keys). */
export function mambaDimsFrom(manifest: Manifest): MambaDims {
  return dimsFromManifest(manifest, "mamba", MAMBA_DIM_KEYS);
}

/**
 * One MambaBlock (mamba.py MambaBlock.forward), no residual — the caller
 * adds it. x: (T, d_model) -> (T, d_model). Records under `prefix`.
 */
function mambaBlock(
  weights: Map<string, T>,
  dims: MambaDims,
  layer: number,
  x: T,
  rec: Recorder | undefined,
): T {
  const { d_state: dState, d_conv: dConv, dt_rank: dtRank } = dims;
  const dInner = dims.expand * dims.d_model;
  const steps = x.shape[0];
  const prefix = `layer${layer}.`;
  const w = (suffix: string): T => getTensor(weights, `blocks.${layer}.${suffix}`);

  // in_proj -> split (x_part, z), x_part first (torch.chunk(2, dim=-1)).
  const xz = linear(x, w("in_proj.weight")); // (T, 2*d_inner)
  rec?.record(`${prefix}in_proj.out`, xz);
  const xPart = sliceCols(xz, 0, dInner);
  const z = sliceCols(xz, dInner, 2 * dInner);

  // Depthwise causal conv along time, then SiLU. conv1d.weight is PyTorch
  // (d_inner, 1, d_conv); squeeze the middle dim to (d_inner, d_conv).
  const convW3 = w("conv1d.weight");
  const convKernels = T.from(convW3.data, [convW3.shape[0], convW3.shape[2]]);
  const xConv = silu(causalDepthwiseConv1d(xPart, convKernels, w("conv1d.bias"), dConv));
  rec?.record(`${prefix}conv.out`, xConv);

  // x_proj -> split (dt, B, C) in that order (torch.split sizes).
  const dbc = linear(xConv, w("x_proj.weight")); // (T, dt_rank + 2*d_state)
  rec?.record(`${prefix}x_proj.out`, dbc);
  const dt = sliceCols(dbc, 0, dtRank);
  const bMat = sliceCols(dbc, dtRank, dtRank + dState);
  const cMat = sliceCols(dbc, dtRank + dState, dtRank + 2 * dState);

  const delta = softplus(linear(dt, w("dt_proj.weight"), w("dt_proj.bias"))); // (T, d_inner)
  rec?.record(`${prefix}delta.out`, delta);

  const A = scale(exp(w("A_log")), -1); // (d_inner, d_state), A = -exp(A_log)
  const D = w("D"); // (d_inner)

  // Selective scan, one token at a time. This is the one place raw indexed
  // loops are used: h is recurrent state, updated in place per step.
  //
  //   h[d,s] = exp(delta[t,d] * A[d,s]) * h[d,s] + delta[t,d] * B[t,s] * x[t,d]
  //   y[t,d] = sum_s h[d,s] * C[t,s] + D[d] * x[t,d]
  const h = T.zeros([dInner, dState]);
  const y = T.zeros([steps, dInner]);
  for (let t = 0; t < steps; t++) {
    for (let d = 0; d < dInner; d++) {
      const deltaTd = delta.data[t * dInner + d];
      const xTd = xConv.data[t * dInner + d];
      for (let s = 0; s < dState; s++) {
        h.data[d * dState + s] =
          Math.exp(deltaTd * A.data[d * dState + s]) * h.data[d * dState + s] +
          deltaTd * bMat.data[t * dState + s] * xTd;
      }
    }
    rec?.record(`${prefix}ssm.h.t${t}`, h); // MapRecorder clones the snapshot
    for (let d = 0; d < dInner; d++) {
      let acc = 0; // f64 accumulation, f32 round on store
      for (let s = 0; s < dState; s++) {
        acc += h.data[d * dState + s] * cMat.data[t * dState + s];
      }
      y.data[t * dInner + d] = acc + D.data[d] * xConv.data[t * dInner + d];
    }
  }
  rec?.record(`${prefix}ssm.out`, y);

  const gated = mul(y, silu(z));
  rec?.record(`${prefix}gate.out`, gated);

  const out = linear(gated, w("out_proj.weight"));
  rec?.record(`${prefix}out_proj.out`, out);
  return out;
}

/**
 * Full forward pass: tokens -> logits (T, vocab), recording the same
 * activation names as mamba.py when `rec` is given.
 */
export function mambaForward(
  weights: Map<string, T>,
  dims: MambaDims,
  tokens: number[],
  rec?: Recorder,
): T {
  let x = embedding(tokens, getTensor(weights, "embedding.weight")); // (T, d_model)
  rec?.record("embed.out", x);

  for (let i = 0; i < dims.n_layer; i++) {
    const normed = rmsnorm(x, getTensor(weights, `norms.${i}.weight`));
    rec?.record(`layer${i}.norm.out`, normed);
    x = add(x, mambaBlock(weights, dims, i, normed, rec));
  }

  x = rmsnorm(x, getTensor(weights, "final_norm.weight"));
  rec?.record("final_norm.out", x);

  const logits = linear(x, getTensor(weights, "lm_head.weight"));
  rec?.record("head.logits", logits);
  return logits;
}
